docker/docker_client.go

```diff
@@ -47,3 +72,53 @@ func NewDockerClient() (dockerClient client.CommonAPIClient, dockerHost string,
 
 	return dockerClient, dockerHost, err
 }
+
+func podmanPresent() bool {
+	_, err := exec.LookPath("podman")
+	return err == nil
+}
+
+// creates a docker client that has its own podman service associated with it
+// the service is shutdown when Close() is called on the client
+func newDockerClientWithPodmanService() (dockerClient client.CommonAPIClient, dockerHost string, err error) {
+	tmpDir, err := os.MkdirTemp("", "func-podman-")
+	if err != nil {
+		return
+	}
+
+	podmanSocket := filepath.Join(tmpDir, "podman.sock")
+	dockerHost = fmt.Sprintf("unix://%s", podmanSocket)
+
+	cmd := exec.Command("podman", "system", "service", dockerHost, "--time=0")
+	err = cmd.Start()
+	if err != nil {
+		return
+	}
+
+	dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithHost(dockerHost), client.WithAPIVersionNegotiation())
+	dockerClient = withPodman{pimpl: dockerClient, stopPodman: func() {
+		_ = cmd.Process.Signal(syscall.SIGTERM)
+		_ = os.RemoveAll(tmpDir)
+	}}
+
+	// give a time to podman to start
+	for i := 0; i < 40; i++ {
+		if _, e := dockerClient.Ping(context.Background()); e == nil {
+			break
+		}
+		time.Sleep(time.Millisecond * 250)
+	}
+
+	return
```

Comment:
Maybe I am misreading this, but if it pings 40 times and never breaks, the value of `err` is not returned because it's using the local variable `e` for ping errors, so it will exit having never been able to successfully ping the service... but also with no error.  Is that the intention?  Or should there be an error returned if not able to ping podman after the 40 tries?